crates/bevy_ecs/src/event/base.rs

```diff
@@ -48,8 +47,41 @@ pub trait Event: Component {
     /// [triggered]: crate::system::Commands::trigger_targets
     /// [`Trigger::propagate`]: crate::observer::Trigger::propagate
     const AUTO_PROPAGATE: bool = false;
+
+    /// Generates the [`ComponentId`] for this event type.
+    ///
+    /// If this type has already been registered,
+    /// this will return the existing [`ComponentId`].
+    ///
+    /// This is used by various dynamically typed observer APIs.
+    fn register_component_id(world: &mut World) -> ComponentId {
```

Comment:
Making these trait methods technically leaves open the ability for users to implement them manually.  If a manual implementation returned, say, `ON_ADD`, that would be bad.  I don't know whether it would actually be unsound, though, and I don't think anyone *would* do that.

Reply:
Left a note!